### ISSUE TITLE
RViz (indigo-devel) catkin unable to find Ogre 1.9 libs (MacOSX10.9.3)

### ISSUE THREAD

**Issue by kpykc** [ros-visualization/rviz#782]
Hello,

i used this gist as a start https://gist.github.com/mikepurvis/9837958 to build ROS.

```
# https://gist.github.com/mikepurvis/9837958

brew tap ros/hydro
brew tap osrf/simulation
brew tap homebrew/versions
brew tap homebrew/science

brew install cmake python libyaml
sudo pip install -U rosdep rosinstall_generator wstool rosinstall
sudo pip install -U pillow
brew install wxpython

rosinstall_generator ros ros_comm robot_model robot_state_publisher diagnostic_msgs octomap rviz \
    --rosdistro indigo --deps --wet-only --tar > indigo.rosinstall
wstool init -j8 src indigo.rosinstall
rosdep install --from-paths src --ignore-src --rosdistro indigo -ry

pushd src
wstool remove orocos_kinematics_dynamics/*
wstool set orocos_kinematics_dynamics --git https://github.com/orocos/orocos_kinematics_dynamics
wstool update orocos_kinematics_dynamics
popd


# http://answers.ros.org/question/94771/building-ros-on-osx-109-solution/
cd <ROS WORKSPACE DIR>

cd src/robot_state_publisher
curl https://github.com/ros/robot_state_publisher/pull/5.patch | patch -p1
cd ..
rm -r build_isolated /robot_state_publisher

cd src/robot_model
curl https://github.com/ros/robot_model/pull/43.patch | patch -p1
cd ..
rm -r build_isolated/kdl_parser

```

I also added to `src/geometry/tf/CMakeLists.txt`:

```
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGTEST_USE_OWN_TR1_TUPLE=1")
```

Then built workspace:

```
 ./src/catkin/bin/catkin_make_isolated --install -DCMAKE_BUILD_TYPE=Release
```

In my case, i also installed manually:

```
brew install opencv --with-qt --with-eigen --with-tbb
brew install lz4
```

Brew has Ogre 1.7 by default, rviz compiles, but failing to start, as it requires Ogre>= 1.8:

```
[ INFO] [1402607169.441462000]: rviz version 1.11.2
[ INFO] [1402607169.441525000]: compiled against OGRE version 1.7.4 (Cthugha)
[ INFO] [1402607169.454891000]: Forcing Stereo OFF
2014-06-12 23:06:09.612 rviz[87898:d0b] invalid drawable
[ INFO] [1402607169.641570000]: Stereo is NOT SUPPORTED
[ INFO] [1402607169.641677000]: OpenGl version: 2.1 (GLSL 1.2).
2014-06-12 23:06:09.751 rviz[87898:d0b] invalid drawable
[ WARN] [1402607169.933062000]: OGRE EXCEPTION(7:InternalErrorException): Cannot find serializer implementation for current version [MeshSerializer_v1.8] in MeshSerializer::importMesh at /tmp/ogre-9kTl/ogre_src_v1-7-4/OgreMain/src/OgreMeshSerializer.cpp (line 118)
libc++abi.dylib: terminating with uncaught exception of type Ogre::InternalErrorException: OGRE EXCEPTION(7:InternalErrorException): Cannot find serializer implementation for current version [MeshSerializer_v1.8] in MeshSerializer::importMesh at /tmp/ogre-9kTl/ogre_src_v1-7-4/OgreMain/src/OgreMeshSerializer.cpp (line 118)
Abort trap: 6
```

I installed Ogre from HEAD:

```
brew install ogre --HEAD
```

But, 

```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
OGRE_OV_RVIZ_LIB_OgreMain
    linked by target "executable" in directory /Users/kp/dev/ros/src/rviz/src/rviz
    linked by target "rviz" in directory /Users/kp/dev/ros/src/rviz/src/rviz
    linked by target "default_plugin" in directory /Users/kp/dev/ros/src/rviz/src/rviz/default_plugin
    linked by target "rviz_image_view" in directory /Users/kp/dev/ros/src/rviz/src/image_view
OGRE_OV_RVIZ_LIB_OgreOverlay
    linked by target "executable" in directory /Users/kp/dev/ros/src/rviz/src/rviz
    linked by target "rviz" in directory /Users/kp/dev/ros/src/rviz/src/rviz
    linked by target "default_plugin" in directory /Users/kp/dev/ros/src/rviz/src/rviz/default_plugin
    linked by target "rviz_image_view" in directory /Users/kp/dev/ros/src/rviz/src/image_view

-- Configuring incomplete, errors occurred!
```

Despite that, as i understood from discussions in rviz issues and source of `CMakeLists.txt` files, support for new Ogre>=1.9 already integrated in `indigo-devel` branch. 

```
> sw_vers
ProductName:    Mac OS X
ProductVersion: 10.9.3
BuildVersion:   13D65

> brew info ogre
ogre: stable 1.7.4, HEAD
http://www.ogre3d.org/
/usr/local/Cellar/ogre/1.7.4 (487 files, 17M)
  Built from source
/usr/local/Cellar/ogre/HEAD (2070 files, 150M) *
  Built from source
From: https://github.com/osrf/homebrew-simulation/commits/master/ogre.rb
==> Dependencies
Build: cmake ✔
Required: boost ✔, doxygen ✔, freeimage ✔, freetype ✔, libzzip ✔, tbb ✔
```

The same both for **rviz-release-release-indigo-rviz-1.11.2-1** retrieved by rosinstall and for current **indigo-devel** branch from git.

**Comment by davetcoleman** [comment]
Thanks for all the hard work on this! I tried your suggested 

> brew install https://raw.githubusercontent.com/hgaiser/homebrew-simulation/master/ogre1.9.rb

As well as changing the branch of rviz to https://github.com/ros-visualization/rviz/tree/osx_ogre_1_9 but I still got the error

> CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
OGRE_OV_RVIZ_LIB_OgreMain
    linked by target "executable" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz
    linked by target "rviz" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz
    linked by target "default_plugin" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz/default_plugin
    linked by target "rviz_image_view" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/image_view
OGRE_OV_RVIZ_LIB_OgreOverlay
    linked by target "executable" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz
    linked by target "rviz" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz
    linked by target "default_plugin" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/rviz/default_plugin
    linked by target "rviz_image_view" in directory /Volumes/Dave/ros/ws_ros/src/rviz/src/image_view

**Comment by hgaiser** [comment]
I believe I 'hacked it up' by adding the following in some CMakeLists.txt where required:

``link_directories(/usr/local/lib)``

Also I had to add some `#ifndef Q_MOC_RUN` `#endif` around some includes to get it to compile (something about a BOOST_JOIN error I believe).

Not sure these are proper fixes, but I had to get it working soon, so I decided not to do a PR with the changes.

I got it running now with the ogre1.9 that you used and the branch you mentioned with the modifications I just mentioned. Works great so far !

**Comment by wjwwood** [comment]
@hgaiser I think you may have stumbled on some solutions we've also arrived at in `rviz`:

> I believe I 'hacked it up' by adding the following in some CMakeLists.txt where required:

There are two issues which are related to this:

- https://github.com/ros-visualization/rviz/issues/828
- https://github.com/ros-visualization/rviz/issues/824
 - This one is blocked on a pull request which I've asked the author to update, but he never got back to it: https://github.com/ros-visualization/rviz/pull/829

I may try to clean up the pr myself and get it merged, I was hopping the original author would do it though.

> Also I had to add some #ifndef Q_MOC_RUN #endif around some includes to get it to compile (something about a BOOST_JOIN error I believe).

This should be fixed in `indigo-devel`, but not released yet: https://github.com/ros-visualization/rviz/pull/826

> @wjwwood, why isn't the rviz ogre 1.9 branch merged into indigo-devel yet?

There were still some outstanding issues and I had not tested it on Linux yet. I'll try to get to that in the short term.

**Comment by jmtatsch** [comment]
I tried the rviz ogre1.9 and indigo-devel branches but I always get a segfault when running it. What is the currently recommended way to compile rviz?

**Comment by jmtatsch** [comment]
Solved my problem by using hgaiser's instead of demmeln's ogre1.9 formula.

**Comment by NikolausDemmel** [comment]
I haven't looked at this in a long time.

After checking out the current status of this a little bit just now, I have the following (OSX 10.9):
- updated to the latest ogre1.9 formula, including the cocoa window patch from hgaiser (I pulled them into my fork as well, which now includes also the latest updates from the osrf main fork). I built the latest v1_9 branch (`brew install ogre1.9 --devel`), not the 1.9.0 release.
- am able to build and run both `indigo-devel` as well as `osx_ogre_1_9` (the latter only afer manually pulling in #826: `curl https://github.com/ros-visualization/rviz/pull/826.patch | patch -p1 --`)
- on `indigo-devel` both camera and image display don't work at all, throwing repeated "invalid drawable" messages when activated
- on the  `osx_ogre_1_9` the image display is working a little bit, with flickering image, and sometimes it stops completely; the camera display causes a segfault
- I did a third try merging the `osx_ogre_1_9` into `indigo-devel`. Here the image display works with flickering as well, but the camera display does not always segfault (it does segfault if activated at the same time as an image display). It does not seem to display anything though.


P.S.: gazebo2 seems to work with that ogre1.9 fomula even without any additional patch as well (https://github.com/osrf/homebrew-simulation/pull/24/files)

**Comment by jmtatsch** [comment]
I can confirm that with Nikolaus's recent changes, rviz works with indigo-devel and brewed python.  However, although I tried to compile everything from scratch against brewed python, I ended up with a system half system half brewed python. E.g. rqt_reconfigure only works when brewed python is unlinked. I would prefer using system python only, but i can't figure out how to install all homebrew deps such as vtk and sip only with system python. sip for instance, won't install brew install sip --without-python. Can someone please give me a hint?

**Comment by NikolausDemmel** [comment]
@jmtatsch: I've only tried with brewed python so far...

**Comment by hawesie** [comment]
@jmtatsch see comment above about gtk+'s dependency on gobject-introspection introducing the brewed python. Remove the brewed python, gobject-introspection and gtk+ and rebuild with `---buid-from-source` and you should be ok.

My additional datapoint: as [above](https://github.com/ros-visualization/rviz/issues/782#issuecomment-74441177), have rviz running with Ogre 1.9 via @NikolausDemmel's formula and the indigo-devel  branch of rviz. gazebo2 compiles and runs with this too. This is with built-in python.

**Comment by jmtatsch** [comment]
Thank you, it has been running stable for me the last couple of days even with brewed python after I cleaned out all python paths and reinstalled home-brew. 
I updated mikepuvis's install script a bit with current ogre/rviz formulas in case anyone is interested:
https://gist.github.com/jmtatsch/3aa268d41ae40c89b35a

**Comment by zurish** [comment]
@kpykc Hello. Is your problem solved? I have basically the problem and did following steps:
1. Installed OGRE1.9, unlinked 1.7.4 and linked 1.9
2. Add
    set(OGRE_OV_RVIZ_LIB_OgreMain "/usr/local/lib/macosx/Release/Ogre.framework/Versions/1.9.0/Ogre")
set(OGRE_OV_RVIZ_LIB_OgreOverlay "/usr/local/lib/macosx/Release/OgreOverlay.framework/Versions/1.9.0/OgreOverlay")
set(OGRE_OV_LIBRARIES_ABS "${OGRE_OV_RVIZ_LIB_OgreMain};${OGRE_OV_RVIZ_LIB_OgreOverlay}")
to CMakeLists.txt of rviz
But when I run rviz, there is still error:
    dyld: Library not loaded: /usr/local/lib/libOgreMain.1.7.4.dylib
    Referenced from: /Users/hxin/ros/install_isolated/lib/rviz/rviz
    Reason: image not found
   Trace/BPT trap: 5
It seems that rviz is still looking for ogre1.7.4. Any idea how to fix this?
Thank you.

**Comment by jmtatsch** [comment]
@zurish try removing the old rviz folders from build, devel, install_isolated  and then compile rviz again

**Comment by zurish** [comment]
@jmtatsch Thanks. Could please explain how to compile it?

**Comment by jmtatsch** [comment]
@zurish don't try the fixes at the very beginning of the thread. If you are new to building ros on OS X, I would recommend to have a look here https://gist.github.com/jmtatsch/3aa268d41ae40c89b35a but be careful with the cleaning part

**Comment by zurish** [comment]
@jmtatsch Thank you. Saw that before and I 'm starting reinstall ros from scratch.

**Comment by zurish** [comment]
@jmtatsch Reinstallation works despite some errors during I was reinstalling it. 
I noticed that rospack can not find all the packages I have in src such as navigation-indigo-revel nor its sub packages like amcl and map_server. Is that normal? Can I still use the unseen packages to rospack? Thanks.

**Comment by jmtatsch** [comment]
Hmm, in my script I assumed the user is using zshell. You may want to source like this if you are using bash:

```source /opt/ros/indigo/setup.bash
source ~/ros_catkin_ws/devel/setup.bash```

**Comment by zurish** [comment]
@jmtatsch Yeah, I tried both zshell and bash, it seems that navigation packages are missing. I tried to install the packages with the command: brew install ros-indigo-navigation
but there is no available formula. How can I install the packages correctly?

**Comment by jmtatsch** [comment]
@zurish just `git clone git@github.com:ros-planning/navigation.git` and then `catkin_make`

**Comment by zurish** [comment]
@jmtatsch Another error occurred during make. A package named orocos-bfl is needed and I cloned it from https://github.com/debrob/orocos-bfl to my src folder. Then I use command:
catkin_make_isolated --pkg orocos-bfl                                       
But output is: 
Packages not found in the workspace: orocos-bfl
What should I do to make catkin identify orocos-bfl? Thank you.

**Comment by NikolausDemmel** [comment]
@jmtatsch: the catkin package is called `bfl`, so you need to `catkin_make_isolated --pkg bfl`. To build a set of packages from source with all dependencies, `rosinstall_generator` is your friend (see the ros from source install instructions). E.g. here for `desktop_full` and `navigation`:

    rosinstall_generator desktop_full navigation --rosdistro indigo --deps --wet-only --tar

**Comment by zurish** [comment]
@NikolausDemmel : Thank you. Yes I see bfl instead of orocos-bfl in the list and I installed bfl. But when I ```catkin_make``` the same error comes
One line of CMakeLists.txt of robot_pose_ekf  package is:
```pkg_check_modules(BFL REQUIRED ocoros-bfl)``` Should I just chang ```ocoros-bfl``` to ```bfl```? 
And please see [my question](http://answers.ros.org/question/204942/orocos-bfl/) at answer.ros.org. Thank you!

**Comment by NikolausDemmel** [comment]
@zurish: this is highly off topic. Let us continue only on answers.ros.org.

**Comment by mikepurvis** [comment]
I'm trying to update my recipe a little with some of the changes from @jmtatsch, but none of these devel formulas for installing Ogre 1.9 seem to persist very long— even with updating the SHA1 hashes, the patch areas move around and cause the formula to fail.

Is there a stable formula for Ogre 1.9 which works from the release tarball rather than VCS branch?

**Comment by wjwwood** [comment]
@mikepurvis I've been working on a cleaned up version of all of this in the last two days (I finally got a new laptop and 10.10). As such I've been trying to collect everything into one Ogre formula and get that formula working with rviz and gazebo5, my version of the Ogre formula is here:

https://github.com/osrf/homebrew-simulation/blob/ogre_1_9/ogre-1.9.rb

By default it works off of the `1.9.0` tag, and unless I'm mistaken that should work.

**Comment by mikepurvis** [comment]
Ah, terrific, thanks very much @wjwwood. That builds Indigo's rviz for me on Yosemite, so I've [applied the update accordingly](https://gist.github.com/mikepurvis/9837958).

Unfortunately, Gazebo is still a no-go— I can build gazebo2 against ogre-1.9, but I get a crash on startup. I definitely had Gazebo 1.9 + Ogre 1.7 operational at one point though. Would it be best to stick with that arrangement, or persist with getting a Gazebo based on Ogre 1.9?

The ideal would be to merge this so that "ogre" is Ogre 1.9, but that would certainly require a clean Gazebo build against it.

**Comment by wjwwood** [comment]
So, rviz works for me, but I'm unable to display an image like @Lenskiy. Maybe it's my test case, but the image does show up in `image_view`, which is annoyingly un-selectable on OS X.

**Comment by wjwwood** [comment]
> Unfortunately, Gazebo is still a no-go

The branch I've linked you to includes a change to `gazebo5` to use the `ogre-1.9` formula instead. I can build and install gazebo and all the gazebo related ROS packages, but gazebo doesn't show a main UI window. I have to iterate with the gazebo guys on that.

My goal it to have `gazebo` and `ogre` installed for people doing Hydro and `gazebo5` and `ogre-1.9` for people doing Indigo/Jade. I'm still working out a strategy for doing that which will likely require multiple homebrew taps.

**Comment by mikepurvis** [comment]
Did anything come of this? I'm currently revamping my install instructions, and wondering if I should be continuing to recommend building Jade and Indigo with Ogre 1.7 and the Hydro meshes?

**Comment by wjwwood** [comment]
I haven't had time to resolve this. There is a related ticket here now:

https://github.com/ros/rosdistro/issues/8912

**Comment by ilidar** [comment]
Well, I've completely removed `ogre` and installed it again using:

`brew install ogre --HEAD`

And it's worked.

FYI: 
1. I'm using `gazebo 6.1` + `ROS indigo`;
2. Tapped only `osrf/simulation`;
3. OS X 10.10.4;

**Comment by grafoteka** [comment]
Hi! i'm trying to update ogre from 1.74 to 1.9 to use it with ros indigo on OS X 10.10.

I'm following the steps on this post, but i'm having some problems.
I tried the next:

MacBook-Pro-de-Jorge:Cellar jdeleon$ brew install https://raw.githubusercontent.com/hgaiser/homebrew-simulation/master/ogre1.9.rb
######################################################################## 100,0%
==> Downloading https://bitbucket.org/sinbad/ogre/get/v1-9-0.tar.gz
Already downloaded: /Library/Caches/Homebrew/ogre1.9-1.9.0.tar.gz
==> Downloading https://gist.githubusercontent.com/NikolausDemmel/2b11d1b49b35cd27a102/raw/3af6b118
Already downloaded: /Library/Caches/Homebrew/ogre1.9--patch-9ad217fc33690f76fd857ba49c3840715d4f3527.diff
==> Downloading https://gist.githubusercontent.com/hgaiser/9ed14de3d776cd34100e/raw/38c7a88cab9067e
Already downloaded: /Library/Caches/Homebrew/ogre1.9--patch-c520d0641183bb275a0b29ef6188353bc2ba6217.patch
==> Patching
patching file CMake/Utils/OgreConfigTargets.cmake
patching file CMakeLists.txt
patching file OgreMain/CMakeLists.txt
patching file RenderSystems/GL/src/OSX/OgreOSXCocoaWindow.mm
patch unexpectedly ends in middle of line
Hunk #1 succeeded at 290 with fuzz 1.
==> cmake -DCMAKE_OSX_ARCHITECTURES='x86_64' -DOGRE_BUILD_LIBS_AS_FRAMEWORKS=OFF -DOGRE_FULL_RPATH:
==> make install
^
3 errors generated.
make[2]: *** [OgreMain/CMakeFiles/OgreMain.dir/src/OSX/OgreConfigDialog.mm.o] Error 1
make[1]: *** [OgreMain/CMakeFiles/OgreMain.dir/all] Error 2
make: *** [all] Error 2

READ THIS: https://git.io/brew-troubleshooting

Also I tried this:
MacBook-Pro-de-Jorge:Cellar jdeleon$ brew install https://raw.githubusercontent.com/NikolausDemmel/homebrew-simulation/ogre-fixes/ogre1.9.rb --devel
######################################################################## 100,0%
==> Downloading https://bitbucket.org/sinbad/ogre/get/v1-9.tar.bz2
######################################################################## 100,0%
Error: SHA1 mismatch
Expected: aa9b0c6d371802e0535515758e88c4893045cd5f
Actual: 545f954b7869f3093f9d7d9ec63638a6aaa8eb4e
Archive: /Library/Caches/Homebrew/ogre1.9-1.9.1-devel.tar.bz2
To retry an incomplete download, remove the file above.

But I have a mismatch error.

Has anyone solve this problem?

**Comment by hamp** [comment]
@grafoteka : had ran into the second problem myself. The SHA1 has to be changed from the "expected" to the "actual" one in the /usr/local/Library/Taps/nikolausdemmel/homebrew-simulation/ogre1.9.rb .

**Comment by wjwwood** [comment]
@grafoteka This requires a change in @hgaiser's homebrew repo: https://github.com/hgaiser/homebrew-simulation

**Comment by mikepurvis** [comment]
FWIW, I don't believe it's possible to build a usable Gazebo 5 from Ogre 1.9 yet on OS X— as of a month or two ago, gzclient segfaults on startup.

What's the motivation for jumping off of Ogre 1.7?

**Comment by hgaiser** [comment]
@mikepurvis, based on what I remember and on the first post it had something to do with the MeshSerializer in Ogre. Based on the first post it seems it requires the serializer from 1.8 or higher. But its been almost a year ago so my memory isn't that clear.

ps. I'll look into ogre not compiling from my homebrew repo.

**Comment by aleksandaratanasov** [comment]
Just so you know this issue is also present on Debian Jessie where both 1.8 and 1.9 dev libs for Ogre are present in the repos and none is detected as installed when building ROS.

**Comment by wjwwood** [comment]
@aleksandaratanasov this issue is for OS X, where the packing is pretty different. If it is not working on Debian, then you should open a new issue. If Debian can install both at the same time, then I'm not surprised that rviz cannot find it correctly since that would have required the Debian packagers to change how the CMake/pkg-config infrastructure is setup in Ogre.

**Comment by aleksandaratanasov** [comment]
I know it's for MacOS. I just added it as a proof that it's actually more wide-spread then one would think reading this issue here.

**Comment by wjwwood** [comment]
@aleksandaratanasov I figured that you knew that, but I was just trying to point out that the packaging is very different between the two systems (in particular OS X's is sort of hacked together), so it's likely to be a different root cause, hence my suggestion to open a new issue.

Thanks.

**Comment by aleksandaratanasov** [comment]
Ah, okay. Yeah, I was going to do that but for some reason it doesn't complain anymore. Go figure...

**Comment by Cescante** [comment]
Hi,
I ran
```
brew install https://raw.githubusercontent.com/hgaiser/homebrew-simulation/master/ogre1.9.rb -v --devel
```
And keep on getting:
```
Undefined symbols for architecture x86_64:
  "std::__1::__vector_base_common<true>::__throw_length_error() const", referenced from:
      void std::__1::vector<boost::shared_ptr<boost::detail::shared_state_base>,
      ...
```
at the step
```
[ 44%] Linking CXX shared library ../lib/macosx/libOgreMain.dylib
cd /tmp/ogre1.920151005-42586-yljmwj/sinbad-ogre-55e89ae88219/build/OgreMain && /usr/local/Cellar/cmake/3.3.2/bin/cmake -E cmake_link_script CMakeFiles/OgreMain.dir/link.txt --verbose=1
/usr/local/Library/ENV/4.3/clang++   -msse -Wall -Winit-self -Wno-overloaded-virtual -Wcast-qual -Wwrite-strings -Wextra -Wno-unused-parameter -Wshadow -Wno-missing-field-initializers -Wno-long-long -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.7 -dynamiclib -Wl,-headerpad_max_install_names -framework IOKit -framework Cocoa -framework Carbon -framework OpenGL -framework CoreVideo -compatibility_version 1.9.0 -current_version 1.9.0 -o ../lib/macosx/libOgreMain.1.9.0.dylib -install_name /tmp/ogre1.920151005-42586-yljmwj/sinbad-ogre-55e89ae88219/build/lib/macosx/libOgreMain.1.9.0.dylib CMakeFiles/OgreMain.dir/src/OgreAlignedAllocator.cpp.o CMakeFiles/OgreMain.dir/src/OgreAnimable.cpp.o CMakeFiles/OgreMain.dir/src/OgreAnimation.cpp.o CMakeFiles/OgreMain.dir/src/OgreAnimationState.cpp.o CMakeFiles/OgreMain.dir/src/OgreAnimationTrack.cpp.o
...
```

I've tried compiling boost 1.58 from source both with and without the --c++11 option with the same result. Looking at various forums it seems that most people who ran into these symbol issues with boost uses a -lc++ compiler flag. Is there some obvious I can do about this? If not I'll go ahead and try to clone and see if the flag fixes my issue.

Thanks.

**Comment by Cescante** [comment]
I ended up modifying the latest ogre's cmake to use libc++ and c++11. Ugly hack but finally have it built and installed:
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -v -stdlib=libc++ -std=c++11 ${OGRE_WARNING_FLAGS}")
```

**Comment by jmtatsch** [comment]
@Cescante so your formula `brew install https://raw.githubusercontent.com/Cescante/homebrew-simulation/master/ogre1.9.rb --devel` worked? Or did you additionally alter the cmakefile?

**Comment by Cescante** [comment]
@jmtatsch The ogre web page says to use CMAKE_CXX_FLAGS = "${CMAKE_CXX_FLAGS} -v -stdlib=libc++ -std=c++11" as part of the cmd args when running cmake. Since we want to do this through homebrew, I couldn't find the right combinations of escaped quotes for the formula to work.

But my machine cached the packages. So I untared it, explicitly added the line mentioned above, repackaged and put back into homebrew's cache directory, updated the checksum in the formula to match the new package, and re-ran brew install.

But then ``` brew install https://raw.githubusercontent.com/NikolausDemmel/homebrew-simulation/ogre-fixes/gazebo2.rb ``` also failed, without any error messages in the log besides one line saying that there were 2 errors. I didn't have any more time to spend on it.

**Comment by mgrrx** [comment]
@Cescante did you install boost with C++11 or without?

**Comment by Cescante** [comment]
@mgrrx With.

**Comment by jmtatsch** [comment]
@Cescante @ffurrer seems to have succeeded in building gazebo6 recently. Maybe he can give us a hint?

**Comment by jmtatsch** [comment]
@Cescante Did you compile Ogre on OS X El Capitan? I managed to get around the libc++ and c++11 issues as you proposed, but I'm stuck after building OgreRTShaderSystem without any meaningful errors https://gist.github.com/jmtatsch/fb8efb7e96e475424697

**Comment by wjwwood** [comment]
@jmtatsch The error from your gist is this:

```
/tmp/ogre1.920151016-16023-xyln3j/sinbad-ogre-55e89ae88219/RenderSystems/GL/src/OSX/OgreOSXWindow.cpp:326:19: error: assigning to 'CGError' from incompatible type 'CGLError' (aka '_CGLError')
            cgErr = CGLSetParameter(mCGLContextObj, kCGLCPSwapInterval, &swapInterval);
                  ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [RenderSystems/GL/CMakeFiles/RenderSystem_GL.dir/src/OSX/OgreOSXWindow.cpp.o] Error 1
make[1]: *** [RenderSystems/GL/CMakeFiles/RenderSystem_GL.dir/all] Error 2
```

I'd recommend asking about it on the Ogre issue tracker.

**Comment by ffurrer** [comment]
I didn't dare to upgrade to El Capitan, yet ;)

**Comment by jmtatsch** [comment]
Seems I was a little too brave then :(
I opened a ticket upstream for my issue
https://ogre3d.atlassian.net/browse/OGRE-506

**Comment by ilidar** [comment]
@jmtatsch Well, what I did is simply casted return value of `CGLSetParameter` function to `CGError` at `OgreOSXWindow.cpp:326`, which helped me to overcome this kind of error.

`cgErr = (CGError)CGLSetParameter(mCGLContextObj, kCGLCPSwapInterval, &swapInterval);`

Yeap, pity, but I have built `ogre1.9` now :dancer: 

P. S.. I've used this formula `https://raw.githubusercontent.com/jmtatsch/homebrew-simulation/master/ogre1.9.rb`, thanks.

**Comment by pgorczak** [comment]
The [CGLError issue](https://ogre3d.atlassian.net/browse/OGRE-506) is fixed. However, on El Capitan (10.11.2), the libc++ issue remains. Changing CMAKE_OSX_DEPLOYMENT_TARGET as suggested in [another ogre issue](https://github.com/OGRECave/ogre/issues/156) did not work for me. The reason might be that I'm not using Xcode. As mentioned before, I had to explicitly add the `-stdlib=libc++` flag.

#### Try as a Patch
I put the one-line patch that fixed it for me [into a gist](https://gist.github.com/pgorczak/e7a23fb799623703282a). You can `brew edit ogre` and add the lines
```
  patch do
    url "https://gist.githubusercontent.com/pgorczak/e7a23fb799623703282a/raw/340aeaec5d21f2e056405590c226d0318ebc026b/ogre_1.9_libc++.diff"
  end if build.head?
```

#### Rebuild Rviz After Patching
*Still working on this. Will fill in later.*

#### Error Message for Reference
```
Last 15 lines from /Users/philipp/Library/Logs/Homebrew/ogre/02.make:
      boost::(anonymous namespace)::thread_proxy(void*) in libboost_thread-mt.a(thread.o)
      boost::thread::join_noexcept() in libboost_thread-mt.a(thread.o)
      boost::thread::do_try_join_until_noexcept(timespec const&, bool&) in libboost_thread-mt.a(thread.o)
      boost::thread::detach() in libboost_thread-mt.a(thread.o)
      boost::thread::interrupt() in libboost_thread-mt.a(thread.o)
      boost::thread::interruption_requested() const in libboost_thread-mt.a(thread.o)
      boost::thread::native_handle() in libboost_thread-mt.a(thread.o)
      ...
  "std::bad_alloc::bad_alloc()", referenced from:
      boost::exception_ptr boost::exception_detail::get_static_exception_object<boost::exception_detail::bad_alloc_>() in libboost_thread-mt.a(thread.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
